src/storage/remote/protocol.c

```diff
@@ -342,16 +497,25 @@ storageRemoteProtocolBlockSize(const String *message)
     ASSERT(message != NULL);
 
     // Create block regular expression if it has not been created yet
-    if (storageRemoteProtocolLocal.memContext == NULL)
+    if (storageRemoteProtocolLocal.blockRegExp == NULL)
     {
-        MEM_CONTEXT_BEGIN(memContextTop())
+        // If the context is not already allocated it means this function is being used locally so allocate it
+        if (storageRemoteProtocolLocal.memContext == NULL)
         {
-            MEM_CONTEXT_NEW_BEGIN("StorageRemoteFileReadLocal")
+            MEM_CONTEXT_BEGIN(memContextTop())
             {
-                storageRemoteProtocolLocal.memContext = memContextCurrent();
-                storageRemoteProtocolLocal.blockRegExp = regExpNew(BLOCK_REG_EXP_STR);
+                MEM_CONTEXT_NEW_BEGIN("StorageRemoteProtocol")
```

Comment:
So in this incarnation of the "StorageRemoteProtocol" we will not have a driver. So in the storageRemoteProtocolLocal has been set up here and then in line 161, the storageRemoteProtocolLocal.driver will be null - but the memContext has already been set up here, so why would we overwrite it in line 167?

Reply:
There are two ways this function can be called -- in the remote *or* the local/main process. So the function can't depend on the context having been created for the driver. Really we are just trying to cache this regexp so it does not need to be created for each block.

I admit that this is a bit convoluted, but since this function will go away when the protocol is binary I'm not very motivated to make it cleaner or pick a better place to store the context.

Reply:
I need to think on this. Take a look at my other 2 replies and I'll revisit tomorrow.

Reply:
Why would the function `storageRemoteFeatureProtocol` at line 142 not look like this, then (I added checking the memory context):
```
void
storageRemoteFeatureProtocol(const VariantList *paramList, ProtocolServer *server)
{
    FUNCTION_LOG_BEGIN(logLevelDebug);
        FUNCTION_LOG_PARAM(VARIANT_LIST, paramList);
        FUNCTION_LOG_PARAM(PROTOCOL_SERVER, server);
    FUNCTION_LOG_END();

    ASSERT(paramList == NULL);
    ASSERT(server != NULL);

    MEM_CONTEXT_TEMP_BEGIN()
    {
        // Get storage based on remote type
        const Storage *storage =
            protocolStorageTypeEnum(cfgOptionStr(cfgOptRemoteType)) == protocolStorageTypeRepo ?
                storageRepoWrite() : storagePgWrite();

        // Store local variables in the server context
        if (storageRemoteProtocolLocal.driver == NULL)
        {
            // If the memoryContext has already been alocatted
            if (storageRemoteProtocolLocal.memContext != NULL)
            {
                MEM_CONTEXT_BEGIN(storageRemoteProtocolLocal.memContext)
                {
                    storageRemoteProtocolLocal.driver = storageDriver(storage);
                }
                MEM_CONTEXT_END();
            }
            else
            {
                MEM_CONTEXT_PRIOR_BEGIN()
                {
                    MEM_CONTEXT_NEW_BEGIN("StorageRemoteProtocol")
                    {
                        storageRemoteProtocolLocal.memContext = memContextCurrent();
                        storageRemoteProtocolLocal.driver = storageDriver(storage);
                    }
                    MEM_CONTEXT_NEW_END();
                }
                MEM_CONTEXT_PRIOR_END();
            }
        }

        // Return storage features
        protocolServerWriteLine(server, jsonFromStr(storagePathP(storage, NULL)));
        protocolServerWriteLine(server, jsonFromUInt64(storageInterface(storage).feature));

        protocolServerResponse(server, NULL);
    }
    MEM_CONTEXT_TEMP_END();

    FUNCTION_TEST_RETURN_VOID();
}
```

Reply:
Mostly because that's more to undo when `storageRemoteProtocolBlockSize()` is removed, which should be in the next few weeks. There might be a simpler way to deal with this, though.

Reply:
534bb04c96bf40894f2 is a simplified version that works fine if we need to leave it in longer than I expect, but removes the interaction with the main context.